SU2_CFD/include/numerics_simd/flow/convection/common.hpp

```diff
@@ -32,37 +32,129 @@
+                                Double relax) {
+  constexpr auto nVarGrad = nVarGrad_ > 0 ? nVarGrad_ : VarType::nVar;
+
+  auto grad_i = gatherVariables<nVarGrad,nDim>(iPoint, gradient);
+  auto grad_j = gatherVariables<nVarGrad,nDim>(jPoint, gradient);
+
   for (size_t iVar = 0; iVar < nVarGrad; ++iVar) {
-    vars(iVar) += scale * dot(grad[iVar], vector_ij);
+    /*--- Centered difference, only needed for U-MUSCL projection ---*/
+    Double delta_ij = 0.0;
+    if (umuscl)
+      delta_ij = V.j.all(iVar) - V.i.all(iVar);
+
+    /*--- U-MUSCL reconstructed variables ---*/
+    const Double proj_i = musclReconstruction(grad_i, delta_ij, vector_ij, iVar, umuscl, kappa, relax);
+    const Double proj_j = musclReconstruction(grad_j, delta_ij, vector_ij, iVar, umuscl, kappa, relax);
+
+    /*--- Apply reconstruction: V_L = V_i + 0.25 * dV_ij^kappa ---*/
+    V.i.all(iVar) += 0.5 * proj_i;
+    V.j.all(iVar) -= 0.5 * proj_j;
   }
 }
 
 /*!
  * \brief Limited reconstruction with point-based limiter.
  */
-template<size_t nVarGrad_ = 0, size_t nVar, size_t nDim, class Limiter_t, class Gradient_t>
+template<size_t nVarGrad_ = 0, size_t nDim, class VarType, class Limiter_t, class Gradient_t>
 FORCEINLINE void musclPointLimited(Int iPoint,
+                                   Int jPoint,
                                    const VectorDbl<nDim>& vector_ij,
-                                   Double scale,
                                    const Limiter_t& limiter,
                                    const Gradient_t& gradient,
-                                   VectorDbl<nVar>& vars) {
-  constexpr auto nVarGrad = nVarGrad_ > 0 ? nVarGrad_ : nVar;
-  auto lim = gatherVariables<nVarGrad>(iPoint, limiter);
-  auto grad = gatherVariables<nVarGrad,nDim>(iPoint, gradient);
+                                   CPair<VarType>& V,
+                                   bool umuscl,
+                                   Double kappa,
+                                   Double relax) {
+  constexpr auto nVarGrad = nVarGrad_ > 0 ? nVarGrad_ : VarType::nVar;
+
+  auto lim_i = gatherVariables<nVarGrad>(iPoint, limiter);
+  auto lim_j = gatherVariables<nVarGrad>(jPoint, limiter);
+
+  auto grad_i = gatherVariables<nVarGrad,nDim>(iPoint, gradient);
+  auto grad_j = gatherVariables<nVarGrad,nDim>(jPoint, gradient);
+
   for (size_t iVar = 0; iVar < nVarGrad; ++iVar) {
-    vars(iVar) += lim(iVar) * scale * dot(grad[iVar], vector_ij);
+    /*--- Centered difference, only needed for U-MUSCL projection ---*/
+    Double delta_ij = 0.0;
+    if (umuscl)
+      delta_ij = V.j.all(iVar) - V.i.all(iVar);
+
+    /*--- U-MUSCL reconstructed variables ---*/
+    const Double proj_i = musclReconstruction(grad_i, delta_ij, vector_ij, iVar, umuscl, kappa, relax);
+    const Double proj_j = musclReconstruction(grad_j, delta_ij, vector_ij, iVar, umuscl, kappa, relax);
+
+    /*--- Apply reconstruction: V_L = V_i + 0.25 * lim * dV_ij^kappa ---*/
```

Comment:
0.25? Same comment about not repeating the formulation. You already organized the code nicely, no need to duplicate via comment :)